File: editorial/chap04/answer_B20.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;
 
int N, M, dp[2009][2009];
string S, T;
 
int main() {
	// 入力
	cin >> S; N = S.size();
	cin >> T; M = T.size();
 
	// 動的計画法
	dp[0][0] = 0;
	for (int i = 0; i <= N; i++) {
		for (int j = 0; j <= M; j++) {
			if (i >= 1 && j >= 1 && S[i - 1] == T[j - 1]) {
				dp[i][j] = min({ dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] });
			}
			else if (i >= 1 && j >= 1) {
				dp[i][j] = min({ dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + 1 });
			}
			else if (i >= 1) {
				dp[i][j] = dp[i - 1][j] + 1;
			}
			else if (j >= 1) {
				dp[i][j] = dp[i][j - 1] + 1;
			}
		}
	}
 
	// 出力
	cout << dp[N][M] << endl;
	return 0;
}

File: editorial/chap04/python/answer_B20.py
# 入力
S = input()
T = input()

# 配列の準備
dp = [ [ None ] * (len(T) + 1) for i in range(len(S) + 1) ]

# 動的計画法
dp[0][0] = 0
for i in range(len(S)+1):
	for j in range(len(T)+1):
		if i>=1 and j>=1 and S[i-1]==T[j-1]:
			dp[i][j] = min(dp[i-1][j]+1, dp[i][j-1]+1, dp[i-1][j-1])
		elif i>=1 and j>=1:
			dp[i][j] = min(dp[i-1][j]+1, dp[i][j-1]+1, dp[i-1][j-1]+1)
		elif i>=1:
			dp[i][j] = dp[i-1][j]+1
		elif j>=1:
			dp[i][j] = dp[i][j-1]+1

# 出力
print(dp[len(S)][len(T)])
